Flaky test in string_test.go

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/kvrocks/issues) and found no similar issues.


### Version

unstable

### Minimal reproduce step

None

### What did you expect to see?

go tests passed

### What did you see instead?

https://github.com/apache/kvrocks/actions/runs/5779002383/job/15660953839#step:12:114

```
--- FAIL: TestString (45.66s)
    server.go:109: 
        	Error Trace:	/Users/runner/work/kvrocks/kvrocks/tests/gocase/unit/type/strings/server.go:109
        	            				/Users/runner/work/kvrocks/kvrocks/tests/gocase/unit/type/strings/server.go:112
        	            				/Users/runner/work/kvrocks/kvrocks/tests/gocase/unit/type/strings/server.go:101
        	            				/Users/runner/work/kvrocks/kvrocks/tests/gocase/unit/type/strings/strings_test.go:757
        	Error:      	An error is expected but got nil.
        	Test:       	TestString
FAIL
exit status 1
FAIL	github.com/apache/kvrocks/tests/gocase/unit/type/strings	45.811s
```


### Anything Else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

Another issue is related to this: https://github.com/apache/kvrocks/actions/runs/5803982834/job/15733157365

```
ok  	github.com/apache/kvrocks/tests/gocase/unit/type/stream	5.497s
panic: Log in goroutine after TestString has completed: 
	Error Trace:	/Users/runner/work/kvrocks/kvrocks/tests/gocase/unit/type/strings/server.go:108
	            				/Users/runner/work/kvrocks/kvrocks/tests/gocase/unit/type/strings/asm_amd64.s:1594
	Error:      	Received unexpected error:
	            	os: process already finished
	Test:       	TestString


goroutine 89 [running]:
testing.(*common).logDepth(0xc000083520, {0xc0003443c0, 0x130}, 0x3)
	/Users/runner/hostedtoolcache/go/1.19.10/x64/src/testing/testing.go:894 +0x4c7
testing.(*common).log(...)
	/Users/runner/hostedtoolcache/go/1.19.10/x64/src/testing/testing.go:876
testing.(*common).Errorf(0xc000083520, {0x10031a4e2?, 0xc000083520?}, {0xc000335880?, 0x100073905?, 0x10054d440?})
	/Users/runner/hostedtoolcache/go/1.19.10/x64/src/testing/testing.go:940 +0x5a
github.com/stretchr/testify/assert.Fail({0x10857da48, 0xc000083520}, {0xc000022880, 0x37}, {0x0, 0x0, 0x0})
	/Users/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.0/assert/assertions.go:264 +0x384
github.com/stretchr/testify/assert.NoError({0x10857da48, 0xc000083520}, {0x10039d740, 0xc0000981a0}, {0x0, 0x0, 0x0})
	/Users/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.0/assert/assertions.go:1392 +0x105
github.com/stretchr/testify/require.NoError({0x10039ec90, 0xc000083520}, {0x10039d740, 0xc0000981a0}, {0x0, 0x0, 0x0})
	/Users/runner/go/pkg/mod/github.com/stretchr/testify@v1.8.0/require/require.go:[126](https://github.com/apache/kvrocks/actions/runs/5803982834/job/15733157365#step:12:127)1 +0x96
github.com/apache/kvrocks/tests/gocase/util.(*KvrocksServer).close.func2()
	/Users/runner/work/kvrocks/kvrocks/tests/gocase/util/server.go:108 +0x8f
created by time.goFunc
	/Users/runner/hostedtoolcache/go/1.19.10/x64/src/time/sleep.go:176 +0x32
exit status 2
FAIL	github.com/apache/kvrocks/tests/gocase/unit/type/strings	45.668s
PASS
ok  	github.com/apache/kvrocks/tests/gocase/unit/type/zset	11.092s
?   	github.com/apache/kvrocks/tests/gocase/util	[no test files]
FAIL
$ /Users/runner/hostedtoolcache/go/1.19.10/x64/bin/go test -timeout=1800s -bench=. ./... -binPath=/Users/runner/work/kvrocks/kvrocks/build/kvrocks -cliPath=redis-cli -workspace=/Users/runner/work/kvrocks/kvrocks/tests/gocase/workspace
Traceback (most recent call last):
  File "/Users/runner/work/kvrocks/kvrocks/./x.py", line 422, in <module>
    args.func(**arg_dict)
  File "/Users/runner/work/kvrocks/kvrocks/./x.py", line 293, in test_go
    run(go, *args, cwd=str(basedir), verbose=True)
  File "/Users/runner/work/kvrocks/kvrocks/./x.py", line 68, in run
    raise RuntimeError(err)
RuntimeError: 
failed to run: ('/Users/runner/hostedtoolcache/go/1.19.10/x64/bin/go', 'test', '-timeout=1800s', '-bench=.', './...', '-binPath=/Users/runner/work/kvrocks/kvrocks/build/kvrocks', '-cliPath=redis-cli', '-workspace=/Users/runner/work/kvrocks/kvrocks/tests/gocase/workspace')
exit with code: 1
```

--- a/tests/gocase/util/server.go
+++ b/tests/gocase/util/server.go
@@ -28,6 +28,7 @@ import (
 	"os/exec"
 	"path/filepath"
 	"regexp"
+	"sync"
 	"syscall"
 	"testing"
 	"time"
@@ -104,11 +105,27 @@ func (s *KvrocksServer) Close() {
 func (s *KvrocksServer) close(keepDir bool) {
 	require.NoError(s.t, s.cmd.Process.Signal(syscall.SIGTERM))
 	f := func(err error) { require.NoError(s.t, err) }
+
+	var wg sync.WaitGroup
 	timer := time.AfterFunc(defaultGracePeriod, func() {
+		defer wg.Done()
+		wg.Add(1)
+
 		require.NoError(s.t, s.cmd.Process.Kill())
-		f = func(err error) { require.EqualError(s.t, err, "signal: killed") }
+		f = func(err error) {
+			// The process may have already exited, so we can't use `require.NoError` here.
+			if err != nil {
+				require.EqualError(s.t, err, "signal: killed")
+			}
+		}
 	})
-	defer timer.Stop()
+
+	defer func() {
+		_ = timer.Stop()
+		// Stop function won't wait for the timer routine if it's already expired,
+		// so we need to wait for it here to prevent panic.
+		wg.Wait()
+	}()
 	f(s.cmd.Wait())
 	s.clean(keepDir)
 }

